File: leetcode/Medium/0054-spiral-matrix/0054-spiral-matrix.py
class Solution:
    def spiralOrder(self, matrix: List[List[int]]) -> List[int]:
        m, n = len(matrix), len(matrix[0])
        answer = []
        now = (0, 0)
        nowd = 0
        d = [(0, 1), (1, 0), (0, -1), (-1, 0)]
        visited = [[False] * n for _ in range(len(matrix))]
        for i in range(m * n):
            answer.append(matrix[now[0]][now[1]])
            visited[now[0]][now[1]] = True
            if not 0 <= now[0] + d[nowd][0] < m or not 0 <= now[1] + d[nowd][1] < n:
                nowd = (nowd + 1) % 4
            elif visited[now[0] + d[nowd][0]][now[1] + d[nowd][1]]:
                nowd = (nowd + 1) % 4
            now = (now[0] + d[nowd][0], now[1] + d[nowd][1])
        return answer

File: leetcode/Medium/0054-spiral-matrix/0054-spiral-matrix.cpp
class Solution {
public:
  vector<int> spiralOrder(vector<vector<int>> &matrix) {
    int m = matrix.size();
    int n = matrix[0].size();
    vector<int> answer = vector<int>();
    pair<int, int> now = {0, 0};
    int nowd = 0;
    vector<pair<int, int>> d = vector<pair<int, int>>();
    d.emplace_back(0, 1);
    d.emplace_back(1, 0);
    d.emplace_back(0, -1);
    d.emplace_back(-1, 0);
    vector<vector<bool>> visited(m, vector<bool>(n, false));

    for (int i = 0; i < m * n; i++) {
      answer.push_back(matrix[now.first][now.second]);
      visited[now.first][now.second] = true;
      int next1 = now.first + d[nowd].first;
      int next2 = now.second + d[nowd].second;
      if (!(0 <= next1 && next1 < m) || !(0 <= next2 && next2 < n)) {
        nowd = (nowd + 1) % 4;
      } else if (visited[next1][next2]) {
        nowd = (nowd + 1) % 4;
      }
      now.first += d[nowd].first;
      now.second += d[nowd].second;
    }
    return answer;
  }
};